### ISSUE TITLE
VMC sets walker count to multiple of number of threads without warning

### ISSUE THREAD

**Issue by prckent** [QMCPACK/qmcpack#1464]
I hit this while testing prime numbers of walkers in the GPU code. The GPU code will correctly run with the number of walkers read from the input. However, the CPU implementation will currently run modulo the number of threads, e.g. walkers=1025 in the GPU code will use 1025 walkers, while a CPU run with 16 threads will use 1024 walkers. It does this e.g. without printing any warnings. What is actually run, e.g. the number of walkers/mpi is correctly printed in all cases, but this change is very subtle.

Possible improvements:
- Print a warning that the code will not do as the user requested
- Abort because the code can not do what the user requested
- Improve the CPU drivers to do as the user requested and print a warning about efficiency.

Are there any opinions on the preferred behavior? I am initially in favor of an abort in the short term and an extension of capabilities in the long term.

**Comment by jtkrogel** [comment]
An abort is too strong for this IMO.  The rounding to number of threads is efficient and hence beneficial to users generally.  

I think we should document this as default behavior and add a user override (round_walkers="no"?) for the less typical case where one must be able to demand an exact number of walkers (like comparative testing).

**Comment by ye-luo** [comment]
I think we need to start with default walker 0;
If not specified in the user input, override it to the number of threads. Otherwise, respect user input.

**Comment by ye-luo** [comment]
walker(should be rename to walker_per_rank) should be hide from user as much as possible.
Users should specify total population size.

**Comment by jtkrogel** [comment]
I agree with this (at least for VMC): "If not specified in the user input, override it to the number of threads. Otherwise, respect user input."

The default (no input) in VMC should be 1 walker per basic processing element (whether thread, gpu, group of threads, or something else).

We need ways to specify either (exclusively) the total number of walkers or the number of walkers per basic processing element (possibly fractional for the latter).

**Comment by prckent** [comment]
Perhaps we fix this when we introduce new keywords that include the normalization explicitly? (total_num_walkers etc.) Then by default, as per @jtkrogel  comment above, we have (1) a standard automatic route, where the user does not request anything and the code chooses appropriately, (2) a route where what the user requests is respected and the code aborts for any unsupported settings.  This makes for simpler code and fewer surprises.

**Comment by zenandrea** [comment]
Hi all,
connected to this discussion, I was wondering what is the ideal number of walkers that I should use per gpu unit. 
I saw in this report 
https://asc.llnl.gov/coral-2-benchmarks/downloads/CORAL2_QMCPACK_vFeb13.pdf
that it is suggested to use 14 walkers per GPU in titan, as it equals "the number of SMs on K20X GPUs".

Do you think that 
walker_per_gpu = SM_per_GPU
is the optimal choice, or should it be something else?

**Comment by prckent** [comment]
@zenandrea  This is notthe optimal choice in general. It is system size dependent and also GPU dependent. Generally one should write a script to try a variety different walker counts, e.g. scaling up in factors of 2. A few steps/blocks of VMC provides a reliable enough timing in my opinion. Kernels have work varying as number of walkers * number of electrons or number of electrons^2, so larger systems need fewer walkers. Hence for small systems one might use hundreds of walkers/gpu, and for very very large systems only a few walkers. Using the largest number of walkers that still fit in GPU memory is usually fastest , but of course all these walkers need to be equilibrated. Smaller counts can be quite close e.g. 1/4 the number of walkers obtains 80% of max performance.
